lighthouse-core/gather/gatherers/source-maps.js

```diff
@@ -0,0 +1,149 @@
+ *
+ * @param {string[]} urls
+ * @return {Promise<string[]>}
+ */
+/* istanbul ignore next */
+async function fetchSourceMaps(urls) {
+  const responses = urls.map(async (url) => {
+    try {
+      const response = await fetch(url);
+      return response.text();
+    } catch (err) {
+      return '!' + err.toString();
+    }
+  });
+  return Promise.all(responses);
+}
+
+/**
+ * @fileoverview Gets JavaScript source maps.
+ */
+class SourceMaps extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+    this._scriptParsedEvents = [];
+  }
+
+  /**
+   * @param {Driver} driver
+   * @param {Array<{url: string, sourceMapURL: string}>} sourceMapsToFetch
+   * @return {Promise<LH.Artifacts.SourceMap[]>}
+   */
+  async fetchSourceMapsInPage(driver, sourceMapsToFetch) {
+    const sourceMapUrlsParam = JSON.stringify(sourceMapsToFetch.map(obj => obj.sourceMapURL));
+    // TODO: change default protocol timeout?
+    // driver.setNextProtocolTimeout(250);
+    /** @type {string[]} */
+    const sourceMapJsons =
+      await driver.evaluateAsync(`(${fetchSourceMaps.toString()})(${sourceMapUrlsParam})`);
+
+    /** @type {LH.Artifacts.SourceMap[]} */
+    const sourceMaps = [];
+    for (const [i, json] of sourceMapJsons.entries()) {
+      const url = sourceMapsToFetch[i].url;
+      if (json.startsWith('!')) {
+        sourceMaps.push({url, error: json.substring(1)});
+      } else {
+        sourceMaps.push(this.parseSourceMapJson(url, json));
+      }
+    }
+    return sourceMaps;
+  }
+
+  /**
+   * @param {string} url
+   * @param {string} json
+   * @return {LH.Artifacts.SourceMap}
+   */
+  parseSourceMapJson(url, json) {
+    try {
+      return {
+        url,
+        map: JSON.parse(json),
+      };
+    } catch (err) {
+      // Without this catch, this silently fails and the gatherer returns an empty object... why no visible error?
+      return {url, error: err.toString()};
+    }
+  }
+
+  /**
+   * @param {LH.Crdp.Debugger.ScriptParsedEvent} event
+   */
+  onScriptParsed(event) {
+    this._scriptParsedEvents.push(event);
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  async beforePass(passContext) {
+    const driver = passContext.driver;
+    driver.on('Debugger.scriptParsed', this.onScriptParsed.bind(this));
+    await driver.sendCommand('Debugger.enable');
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['SourceMaps']>}
+   */
+  async afterPass(passContext) {
+    const driver = passContext.driver;
+
+    driver.off('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.disable');
+
+    /** @type {LH.Artifacts.SourceMap[]} */
+    const sourceMaps = [];
+
+    /** @type {Array<{url: string, sourceMapURL: string}>} */
+    const toFetch = [];
+    for (const event of this._scriptParsedEvents) {
+      if (event.sourceMapURL) {
```

Comment:
super nit: maybe we could filter these out in our listener to avoid the bit of overhead here? that or `if (!...) continue` it's mainly the nesting I'm nitting on :)

Reply:
done

I tried changing the type to `Array<LH.Crdp.Debugger.ScriptParsedEvent & {sourceMapURL: string}>` as well, but got this unexpected TS error:

![image](https://user-images.githubusercontent.com/4071474/58837467-888c3700-8610-11e9-971b-1fc1d1d1e030.png)